### PR TITLE
[SERVICE-340] Offline Mode

### PR DESCRIPTION
**Pull request by ryanpetersonOF**
Adds offline mode to the provider!  We cache the provider js assets, as well as all tabstrip assets.  Requires one "online" launch but after that you can run it offline.  We use the network first recipe, so if you are online you'll use the network, otherwise you'll load from the cache.   The cache uses the layouts version from the package.json so it will be updated with each new release.

### CODE REVIEW

**Comment by finos-admin** [comment]
Thank you for your pull request and welcome to our community! We require contributors to sign a [Contributor License Agreement](https://finosfoundation.atlassian.net/wiki/spaces/FINOS/pages/75530375/Legal+Requirements#LegalRequirements-ContributorLicenseAgreement) and we don't seem to have CLAs on file for these contributors to the Pull Request: (@imranopenfin). In order for your PR to be reviewed and merged, please follow the directions at the link above.

Project team: please **do not merge this Pull Request** until Foundation staff have confirmed that a CLA is in place for the new contributor(s) listed above.

If there are any questions, please don't hesitate to [get in touch with our Infrastructure Support team](https://finosfoundation.atlassian.net/wiki/spaces/FINOS/pages/77955298/Getting+in+Touch#GettinginTouch-InfrastructureSupport). /CC @finos-admin

**Comment by finos-admin** [comment]
Thank you for your pull request and welcome to our community! We require contributors to sign a [Contributor License Agreement](https://finosfoundation.atlassian.net/wiki/spaces/FINOS/pages/75530375/Legal+Requirements#LegalRequirements-ContributorLicenseAgreement) and we don't seem to have CLAs on file for these contributors to the Pull Request: (@imranopenfin). In order for your PR to be reviewed and merged, please follow the directions at the link above.

Project team: please **do not merge this Pull Request** until Foundation staff have confirmed that a CLA is in place for the new contributor(s) listed above.

If there are any questions, please don't hesitate to [get in touch with our Infrastructure Support team](https://finosfoundation.atlassian.net/wiki/spaces/FINOS/pages/77955298/Getting+in+Touch#GettinginTouch-InfrastructureSupport). /CC @finos-admin

**Comment by finos-admin** [comment]
Thank you for your pull request and welcome to our community! We require contributors to sign a [Contributor License Agreement](https://finosfoundation.atlassian.net/wiki/spaces/FINOS/pages/75530375/Legal+Requirements#LegalRequirements-ContributorLicenseAgreement) and we don't seem to have CLAs on file for these contributors to the Pull Request: (@imranopenfin). In order for your PR to be reviewed and merged, please follow the directions at the link above.

Project team: please **do not merge this Pull Request** until Foundation staff have confirmed that a CLA is in place for the new contributor(s) listed above.

If there are any questions, please don't hesitate to [get in touch with our Infrastructure Support team](https://finosfoundation.atlassian.net/wiki/spaces/FINOS/pages/77955298/Getting+in+Touch#GettinginTouch-InfrastructureSupport). /CC @finos-admin

File: res/provider/provider.html [hunk 1]
```diff
@@ -12,6 +12,12 @@
   <h3>Layouts Service Provider</h3>
 
   <button onclick="fin.System.showDeveloperTools({uuid: 'layouts-service', name: 'layouts-service'})">Show Debugger</button>
+
+  <script>
+  if ('serviceWorker' in navigator) {
+    navigator.serviceWorker.register('./sw.js', {scope: './'});
+  }
```
Comment (pjbroadbent) [inline res/provider/provider.html]: Can this be done in `main.ts`?

File: src/provider/ServiceWorker.js [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+const cacheFiles = [
+    './provider.html',
+    './main-bundle.js',
+    './tabStrip-bundle.js',
+    './tabbing/tabstrip/tabstrip.html',
+    './tabbing/tabstrip/css/tabs.css',
+    './tabbing/tabstrip/css/image/close.png',
+    './tabbing/tabstrip/css/image/maximise.png',
+    './tabbing/tabstrip/css/image/minimize.png',
+    './tabbing/tabstrip/css/image/restore.png'
```
Comment (pjbroadbent) [inline src/provider/ServiceWorker.js]: Does this allow glob patterns, or does it have to be a "full" file list?

`app.json` isn't in this list, assuming the RVM & `offlineAccess` takes care of that?

Comment (ryanpetersonOF) [reply src/provider/ServiceWorker.js]: No patterns allowed here.  This is the list of precached files.  Chromium wont go out and do any searching, unfortunately.  App.json is handled by the RVM

File: res/provider/provider.html (hunk 1 above)
Comment (ryanpetersonOF) [reply res/provider/provider.html]: Fixed

**Comment by finos-admin** [comment]
Thank you for your pull request and welcome to our community! We require contributors to sign a [Contributor License Agreement](https://finosfoundation.atlassian.net/wiki/spaces/FINOS/pages/75530375/Legal+Requirements#LegalRequirements-ContributorLicenseAgreement) and we don't seem to have CLAs on file for these contributors to the Pull Request: (@imranopenfin). In order for your PR to be reviewed and merged, please follow the directions at the link above.

Project team: please **do not merge this Pull Request** until Foundation staff have confirmed that a CLA is in place for the new contributor(s) listed above.

If there are any questions, please don't hesitate to [get in touch with our Infrastructure Support team](https://finosfoundation.atlassian.net/wiki/spaces/FINOS/pages/77955298/Getting+in+Touch#GettinginTouch-InfrastructureSupport). /CC @finos-admin

**Comment by MichaelMCoates** [comment]
Is there any way for us to run a basic set of tests on the service worker?

**Comment by ryanpetersonOF** [comment]
@MichaelMCoates  Not really an easy way to test it.  We would have to kill the 'server' which is also running the tests.  Its easily tested manually so I think we should opt for that

**Pull request merged**